src/Oscoin/Storage/BlockTree/Internal.hs

```diff
@@ -0,0 +1,41 @@
+-- | An opaque 'BlockTree', composed by its public and private API. Its main
+-- responsibilities are:
+--
+-- 1. Storing main blocks.
+-- 2. Storing orphan blocks.
+-- 3. Performing chain selection.
+-- 4. Perform full-chain validation.
+
+module Oscoin.Storage.BlockTree.Internal where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Blockchain (TxLookup)
+import           Oscoin.Crypto.Blockchain.Block
+import           Oscoin.Crypto.Hash (Hashed)
+import           Oscoin.Telemetry.Events (NotableEvent)
+import           Oscoin.Time.Chrono (NewestFirst, OldestFirst)
+
+type BlockTree c tx s m = (BlockTreeReader c tx s m, BlockTreeWriter c tx s m)
+
+data BlockTreeReader c tx s m = BlockTreeReader
+    { getGenesisBlock :: m (Block c tx (Sealed c s))
+    -- ^ Get the genesis block.
+    , lookupBlock     :: BlockHash c -> m (Maybe (Block c tx (Sealed c s)))
+    -- ^ Lookups a 'Block' in the tree when given its hash.
+    , lookupTx        :: Hashed c tx -> m (Maybe (TxLookup c tx))
+    -- ^ Lookups a transaction by its hash.
+    , getBlocksByDepth :: Depth -> m (NewestFirst [] (Block c tx (Sealed c s)))
+    -- ^ Returns the last N blocks, given the 'Depth'.
+    , getBlocksByParentHash :: BlockHash c -> m (NewestFirst [] (Block c tx (Sealed c s)))
+    -- ^ Returns the blocks starting from the input parent hash.
```

Comment:
This is a bot hard to understand. I think this function is supposed to return all descendants of a given block. How does this play with the tree? Do we return an empty list if the block is not in the best chain? Do we always return the best chain of descendants?

Reply:
Right -- we discussed a bit these names with @cloudhead in some previous incantations of this work, and we converged with these 2 names. I suppose the ambiguity would be solved if once again we treat our `BlockTreeReader` as a "main chain oracle" so that you don't have branch in those cases but ideally just a liner sequence of blocks (also known as a `Blockchain` 😁 )